src/main/java/org/opensearch/geospatial/ip2geo/processor/Ip2GeoProcessor.java

```diff
@@ -0,0 +1,430 @@
+                    ingestDocument.setFieldValue(targetField, DATA_EXPIRED);
+                    handler.accept(ingestDocument, null);
+                    return;
+                }
+
+                GeoIpDataHelper.getGeoData(client, datasource.currentIndexName(), ip, new ActionListener<>() {
+                    @Override
+                    public void onResponse(final Map<String, Object> stringObjectMap) {
+                        cache.put(ip, datasourceName, stringObjectMap);
+                        if (!stringObjectMap.isEmpty()) {
+                            ingestDocument.setFieldValue(targetField, filteredGeoData(stringObjectMap, ip));
+                        }
+                        handler.accept(ingestDocument, null);
+                    }
+
+                    @Override
+                    public void onFailure(final Exception e) {
+                        handler.accept(null, e);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(final Exception e) {
+                handler.accept(null, e);
+            }
+        });
+    }
+
+    /**
+     * Handle multiple ips
+     *
+     * @param ingestDocument the document
+     * @param handler the handler
+     * @param ips the ip list
+     */
+    private void executeInternal(
+        final IngestDocument ingestDocument,
+        final BiConsumer<IngestDocument, Exception> handler,
+        final List<?> ips
+    ) {
+        Map<String, Map<String, Object>> data = new HashMap<>();
+        for (Object ip : ips) {
+            if (ip instanceof String == false) {
+                throw new IllegalArgumentException("array in field [" + field + "] should only contain strings");
+            }
+            String ipAddr = (String) ip;
+            data.put(ipAddr, cache.get(ipAddr, datasourceName));
+        }
+        List<String> ipList = (List<String>) ips;
+        DatasourceHelper.getDatasource(client, datasourceName, new ActionListener<>() {
+            @Override
+            public void onResponse(final Datasource datasource) {
+                if (datasource == null) {
+                    handler.accept(null, new IllegalStateException("datasource does not exist"));
+                    return;
+                }
+
+                if (datasource.isExpired()) {
+                    ingestDocument.setFieldValue(targetField, DATA_EXPIRED);
+                    handler.accept(ingestDocument, null);
+                    return;
+                }
+                GeoIpDataHelper.getGeoData(
+                    client,
+                    datasource.currentIndexName(),
+                    ipList.iterator(),
+                    maxBundleSize,
+                    maxConcurrentSearches,
+                    firstOnly,
+                    data,
+                    new ActionListener<>() {
+                        @Override
+                        public void onResponse(final Object obj) {
```

Comment:
can we combine the logic of converting list to geo data and single element in 1 common place. I see a lot of commonalities.

I think there is an opportunity.

Reply:
I removed some duplicated code in my next PR which is ready. Combining them are not straight forward.